File: backend/models/Job.js
import mongoose from 'mongoose';

// ─── Job Model ───────────────────────────────────────
// Represents a job posting created by a recruiter.
// Candidates can view and apply to these postings.

const jobSchema = new mongoose.Schema({
  // The recruiter who created this job posting
  postedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Job must have a recruiter'],
  },
  title: {
    type: String,
    required: [true, 'Job title is required'],
    trim: true,
  },
  description: {
    type: String,
    required: [true, 'Job description is required'],
  },
  company: {
    type: String,
    required: [true, 'Company name is required'],
    trim: true,
  },
  location: {
    type: String,
    default: 'Remote',
    trim: true,
  },
  type: {
    type: String,
    enum: ['Full-Time', 'Part-Time', 'Internship', 'Contract'],
    default: 'Full-Time',
  },
  // Required skills for the position
  skills: {
    type: [String],
    default: [],
  },
  salary: {
    type: String,
    default: null, // e.g. "₹6-10 LPA" or "$80k-$120k"
  },
  // How many positions are open
  openings: {
    type: Number,
    default: 1,
    min: 1,
  },
  status: {
    type: String,
    enum: ['open', 'closed', 'draft'],
    default: 'open',
  },
  // Deadline for applications
  deadline: {
    type: Date,
    default: null,
  },
}, {
  timestamps: true,
});

// Index for efficient queries
jobSchema.index({ postedBy: 1, status: 1 });
jobSchema.index({ company: 1 });
jobSchema.index({ status: 1, createdAt: -1 });

const Job = mongoose.model('Job', jobSchema);
export default Job;

File: backend-spring/src/main/java/com/learn2hire/model/Job.java
package com.learn2hire.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "jobs")
@CompoundIndexes({
    @CompoundIndex(name = "postedBy_status", def = "{'postedBy': 1, 'status': 1}"),
    @CompoundIndex(name = "company_idx", def = "{'company': 1}"),
    @CompoundIndex(name = "status_createdAt", def = "{'status': 1, 'createdAt': -1}")
})
public class Job {

    @Id
    @JsonProperty("_id")
    private String id;

    private String postedBy; // User ID reference

    private String title;

    private String description;

    private String company;

    @Builder.Default
    private String location = "Remote";

    @Builder.Default
    private String type = "Full-Time"; // Full-Time, Part-Time, Internship, Contract

    @Builder.Default
    private List<String> skills = new ArrayList<>();

    private String salary;

    @Builder.Default
    private Integer openings = 1;

    @Builder.Default
    private String status = "open"; // open, closed, draft

    private Date deadline;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}
